github/provider_test.go

```diff
@@ -40,301 +34,196 @@ func init() {
-
-	username := "hashibot"
-	resource.Test(t, resource.TestCase{
-		PreCheck: func() {
-			testAccPreCheck(t)
-		},
-		Providers: testAccProviders,
-		Steps: []resource.TestStep{
-			{
-				Config:      providerConfig + testAccCheckGithubUserDataSourceConfig(username),
-				ExpectError: regexp.MustCompile("x509: certificate is valid for untrusted, not localhost"),
-			},
-		},
-	})
-}
+		url, closeFunc := githubTLSApiMock(port, certFile, keyFile, t)
+		defer func() {
+			err := closeFunc()
+			if err != nil {
+				t.Fatal(err)
+			}
+		}()
 
-func githubTLSApiMock(port, certFile, keyFile string, t *testing.T) (string, func() error) {
-	mux := http.NewServeMux()
+		oldBaseUrl := os.Getenv("GITHUB_BASE_URL")
+		defer os.Setenv("GITHUB_BASE_URL", oldBaseUrl)
 
-	userPattern := "/v3/users/hashibot"
-	orgPattern := "/v3/orgs/" + testOwner
+		// Point provider to mock API with self-signed cert
+		os.Setenv("GITHUB_BASE_URL", url)
 
-	mux.HandleFunc(userPattern, testRespondJson(userResponseBody))
-	mux.HandleFunc(userPattern+"/gpg_keys", testRespondJson(gpgKeysResponseBody))
-	mux.HandleFunc(userPattern+"/keys", testRespondJson(keysResponseBody))
-	mux.HandleFunc(orgPattern, testRespondJson(orgResponseBody(port)))
+		providerConfig := `provider "github" {}`
 
-	server := &http.Server{
-		Addr:    ":" + port,
-		Handler: mux,
-	}
+		username := "hashibot"
+		resource.Test(t, resource.TestCase{
+			Providers: testAccProviders,
+			Steps: []resource.TestStep{
+				{
+					Config:      providerConfig + testAccCheckGithubUserDataSourceConfig(username),
+					ExpectError: regexp.MustCompile("x509: certificate is valid for untrusted, not localhost"),
+				},
+			},
+		})
 
-	// nolint: errcheck
-	go server.ListenAndServeTLS(certFile, keyFile)
+	})
 
-	return "https://localhost:" + port + "/", server.Close
-}
+	t.Run("can be configured to run anonymously", func(t *testing.T) {
+
+		anonymousConfiguration := `
+			provider "github" {}
+			data "github_ip_ranges" "test" {}
+		`
+		anonymousCheck := resource.ComposeTestCheckFunc(
+			resource.TestCheckResourceAttrSet("data.github_ip_ranges.test", "hooks.#"),
+			resource.TestCheckResourceAttrSet("data.github_ip_ranges.test", "git.#"),
+			resource.TestCheckResourceAttrSet("data.github_ip_ranges.test", "pages.#"),
+			resource.TestCheckResourceAttrSet("data.github_ip_ranges.test", "importer.#"),
+		)
+
+		resource.Test(t, resource.TestCase{
+			PreCheck:  func() {},
+			Providers: testAccProviders,
+			Steps: []resource.TestStep{
+				{
+					Config: anonymousConfiguration,
+					Check:  anonymousCheck,
+				},
+			},
+		})
+	})
 
-func testRespondJson(responseBody string) func(http.ResponseWriter, *http.Request) {
-	return func(w http.ResponseWriter, req *http.Request) {
-		w.Header().Set("Content-Type", "application/json")
-		if _, err := w.Write([]byte(responseBody)); err != nil {
-			return
-		}
-	}
-}
+	t.Run("can be configured with an individual account", func(t *testing.T) {
```

Comment:
ah i missed this on first review :/ side-effect of adding the `ConflictsWith` property in the provider is that this and the subsequent test can't be run at the same time without 1 or the other failing since only one of the env vars can be set so i think instead of eventually failing at `testAccPreCheckEnvironment` the env var could be checked here in the beginning of the method and call t.Skip()

Reply:
Another approach could be to unset the conflicting variable ahead of each test and `defer` re-instating the variable after the test has run.  This pattern is used in the `insecure` mode tests in the past.  I will experiment with both options and return with a commit.

Reply:
aaa1d86187961252571c3d344acd74d8df746b85 takes a different approach and removes the `ConflictsWith` feature for this use case.  It seems like the cleanest solution for the CI use case, but we likely confuse someone down the line if the docs are not clear.  What do you think?